### PR TITLE
Add CSV export to the evaluation tool

### PR DESCRIPTION
**Pull request by kdaun**
- adds `covered_distance` to the ground truth relations
- the flags `write_relation_metrics` and `relation_metrics_filename` control whether and where to write the relations metrics as comma-separated values

### CODE REVIEW

File: cartographer/ground_truth/autogenerate_ground_truth_main.cc [hunk 1]
```diff
@@ -121,9 +121,10 @@ proto::GroundTruth GenerateGroundTruth(
         submap_to_node_index.at(constraint.submap_id().submap_index());
 
     // Covered distance between the two should not be too small.
-    if (std::abs(covered_distance.at(matched_node) -
-                 covered_distance.at(representative_node)) <
-        min_covered_distance) {
+    double covered_distance_in_constraint =
+        covered_distance.at(matched_node) -
```
Comment (gaschler) [inline cartographer/ground_truth/autogenerate_ground_truth_main.cc]: Why do you drop the abs?

File: cartographer/ground_truth/compute_relations_metrics_main.cc [hunk 2]
```diff
@@ -184,7 +241,16 @@ int main(int argc, char** argv) {
     google::ShowUsageWithFlagsRestrict(argv[0], "compute_relations_metrics");
     return EXIT_FAILURE;
   }
-  ::cartographer::ground_truth::Run(FLAGS_pose_graph_filename,
-                                    FLAGS_relations_filename,
-                                    FLAGS_read_text_file_with_unix_timestamps);
+  std::string relation_metrics_filename;
+  if (FLAGS_relation_metrics_filename != "") {
+    relation_metrics_filename = FLAGS_relation_metrics_filename;
+  } else {
```
Comment (gaschler) [inline cartographer/ground_truth/compute_relations_metrics_main.cc]: make this an if, drop the one above

File: cartographer/ground_truth/compute_relations_metrics_main.cc [hunk 3]
```diff
@@ -43,6 +43,11 @@ DEFINE_string(relations_filename, "",
 DEFINE_bool(read_text_file_with_unix_timestamps, false,
             "Enable support for the relations text files as in the paper. "
             "Default is to read from a GroundTruth proto file.");
+DEFINE_bool(write_relation_metrics, false,
+            "Enable exporting relation metrics as comma-separated values.");
+DEFINE_string(relation_metrics_filename, "",
```
Comment (gaschler) [inline cartographer/ground_truth/compute_relations_metrics_main.cc]: Do you need this feature to set non-default file paths?

File: cartographer/ground_truth/compute_relations_metrics_main.cc [hunk 4]
```diff
@@ -43,6 +43,11 @@ DEFINE_string(relations_filename, "",
...
+              "Export target for relation metrics. Defaults to "
```
Comment (gaschler) [inline cartographer/ground_truth/compute_relations_metrics_main.cc]: Replace target by file path

File: cartographer/ground_truth/compute_relations_metrics_main.cc [hunk 5]
```diff
@@ -109,6 +114,52 @@ std::string StatisticsString(const std::vector<Error>& errors) {
...
 
+void WriteRelationMetricsToFile(const std::vector<Error>& errors,
+                                const proto::GroundTruth& ground_truth,
+                                const std::string& relation_metrics_filename) {
+  std::vector<double> translational_errors;
+  std::vector<double> squared_translational_errors;
+  std::vector<double> rotational_errors_degrees;
+  std::vector<double> squared_rotational_errors_degrees;
+  CHECK_EQ(errors.size(), ground_truth.relation_size());
+  size_t num_errors = errors.size();
```
Comment (gaschler) [inline cartographer/ground_truth/compute_relations_metrics_main.cc]: drop this, just use size()

File: cartographer/ground_truth/compute_relations_metrics_main.cc [hunk 6]
```diff
@@ -109,6 +114,52 @@ std::string StatisticsString(const std::vector<Error>& errors) {
          MeanAndStdDevString(squared_rotational_errors_degrees) + " deg^2\n";
 }
 
+void WriteRelationMetricsToFile(const std::vector<Error>& errors,
+                                const proto::GroundTruth& ground_truth,
+                                const std::string& relation_metrics_filename) {
+  std::vector<double> translational_errors;
```
Comment (gaschler) [inline cartographer/ground_truth/compute_relations_metrics_main.cc]: drop all these you are not using

**Comment by kdaun** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user kdaun. Merge queue now has a length of 1.

**Pull request merged**